src/ir/match.h

```diff
@@ -0,0 +1,291 @@
+/*
+ * Copyright 2020 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//
+// match.h: Convenience structs for matching and transforming Binaryen IR
+// patterns. Most matchers take an optional pointer to an Expression* variable
+// that will be set equal to the expression they match. When building a pattern,
+// the matchers will instead insert the optional Expression* at their location
+// in the pattern.
+//
+
+#ifndef wasm_ir_match_h
+#define wasm_ir_match_h
+
+#include "wasm.h"
+
+namespace wasm {
+
+namespace Match {
+
+// The available matchers are:
+//
+//  i32(x), i64(x), f32(x), f64(x)
+//
+//    Match constants of the exact corresponding type. Do not support building.
+//
+//  ival(x), fval(x)
+//
+//    Match any integer constant or any floating point constant. Do not support
+//    building.
+//
+//  constant([e])
+//
+//    Matches any Const expression and binds it to `e`, if provided.
+//
+//  any([e])
+//
+//    Matches any expression and binds it to `e`, if provided.
+//
+//  unary([e], op, val)
+//
+//    Matches Unary expressions implementing UnaryOp `op` whose operand matches
+//    `val`. Binds the matched expression to `e`, if provided.
+//
+//  binary([e], op, left, right)
+//
+//    Matches Binary expressions implementing BinaryOp `op` whose operands match
+//    `left` and `right`. Binds the matched expression to `e`, if provided.
+//
+//  select([e], ifTrue, ifFalse, cond)
+//
+//    Matches Select expressions whose operands match `ifTrue`, `ifFalse`, and
+//    `cond`. Binds the matched expression to `e`, if provided.
+//
+
+// The main entrypoint for matching.
+template<class Matcher> bool matches(Expression* expr, Matcher matcher) {
+  return matcher.matches(expr);
+}
+
+// The main entrypoint for building. The top-level matcher must point to an
+// expression, which will be modified in place by inserting bound expressions at
+// the locations corresponding to their position in the pattern. Matchers in the
+// pattern that do not have bound expressions are assumed to match the output.
+// Returns the top-level modified expression.
+template<class Matcher> Expression* build(Matcher matcher) {
+  Expression* expr = nullptr;
+  matcher.build(&expr);
+  return expr;
+}
+
+// Matchers
+
+// TODO: Make this a template once we have C++20 (needs non-type class template
+// arguments)
```

Comment:
I didn't know much about this and searched a little, and it seems like non-type class template arguments are currently supported but with some limitations. What does C++20 additionally allow you want to use here?

Reply:
Specifically, I wanted to be able to pass a lambda encapsulating the `c->type == Type::i32` logic, but actually I'm not sure whether C++20 allows that or not.